# VCR0020: Result value accessed without success check

| Property    | Value                          |
|-------------|--------------------------------|
| **Rule ID** | VCR0020                        |
| **Category**| Usage                          |
| **Severity**| Warning                        |
| **Enabled** | Yes                            |

## Description

Accessing `Result<T>.Value` without first checking `IsSuccess` may lead to using a default/null value when the operation failed. Use `Match`, `Switch`, or check `IsSuccess` before accessing `Value`.

## Examples

### Violating code

```csharp
var result = GetUser(id);
// VCR0020: Access to 'Value' on 'result' without checking 'IsSuccess'
var name = result.Value.Name;

// Also flagged: method call chain without variable
var name = GetUser(id).Value.Name;
```

### Fixed code

```csharp
// Option 1: Match (recommended)
var name = result.Match(
    user => user.Name,
    error => "Unknown");

// Option 2: IsSuccess guard with early return
if (result.IsFailure)
    return Error.NotFoundError("User not found");
var name = result.Value.Name;

// Option 3: IsSuccess guard in same scope
if (result.IsSuccess)
{
    var name = result.Value.Name;
    // ...
}

// Option 4: GetValueOrThrow (explicit exception boundary)
var name = result.GetValueOrThrow().Name;
```

## Recognized guard patterns

The analyzer recognizes the following patterns as valid guards:

- `if (result.IsSuccess) { ... result.Value ... }` — success branch of conditional
- `if (result.IsFailure) return;` — early-return guard before `.Value` access
- `if (result.IsFailure) throw ...;` — early-throw guard
- `if (result.IsFailure) { result = Result<T>.Success(...); }` — reassignment guard
- `result.IsSuccess && result.Value ...` — short-circuit guard
- `Assert.True(result.IsSuccess)` / `Assert.False(result.IsFailure)` — xUnit assert guards
- `Assert.IsTrue(result.IsSuccess)` / `Assert.IsFalse(result.IsFailure)` — NUnit / MSTest assert guards
- `Assert.That(result.IsSuccess, ...)` — NUnit constraint-based assert guard (see note below)
- `result.IsSuccess.Should().BeTrue()` / `result.IsFailure.Should().BeFalse()` — FluentAssertions guards

**Note:** `Assert.That` recognition checks only the first argument. The NUnit constraint (second argument) is not inspected — `Assert.That(result.IsSuccess, Is.False)` would be incorrectly treated as a success guard. In practice this is not an issue because `Assert.That(x.IsFailure, Is.True)` is the idiomatic way to assert failure.

## When to suppress

Suppress in test projects where assertion patterns are not recognized:

```editorconfig
# .editorconfig (test projects)
[*Tests/**/*.cs]
dotnet_diagnostic.VCR0020.severity = suggestion
```

## See also

- [VCR0010 — Result must be consumed](VCR0010.md)
- [VCR0040 — GetValueOrThrow defeats Result pattern](VCR0040.md)
