# VCR0050: Failure created with Error.None

| Property    | Value                          |
|-------------|--------------------------------|
| **Rule ID** | VCR0050                        |
| **Category**| Usage                          |
| **Severity**| Error                          |
| **Enabled** | Yes                            |

## Description

`Result.Failure(Error.None)` creates a failure result without an actual error, which is a semantic contradiction. A failure must always carry a meaningful error explaining what went wrong. This is always a bug.

## Examples

### Violating code

```csharp
// VCR0050: Creating a Failure result with Error.None is a bug
return Result.Failure(Error.None);

// Also flagged with generic version
return Result<User>.Failure(Error.None);
```

### Fixed code

```csharp
// Option 1: Use a specific error
return Result.Failure(Error.ValidationError("Invalid input"));

// Option 2: If success was intended
return Result.Success();

// Option 3: Use UnexpectedError for unknown failures
return Result.Failure(Error.UnexpectedError("Operation failed"));
```

## Why this is an error (not warning)

`Error.None` represents the absence of an error. Combining it with `Failure` creates a contradictory state: a failed result that claims nothing went wrong. This will confuse any downstream code that examines the error.

## When to suppress

This rule should never be suppressed — `Failure(Error.None)` is always a bug.

## See also

- [Error Types](../error-types.md)
- [VCR0010 — Result must be consumed](VCR0010.md)
